fem/transfer.cpp

```diff
@@ -522,6 +522,439 @@ const Operator &L2ProjectionGridTransfer::BackwardOperator()
+
+void L2ProjectionH1GridTransfer::L2ProjectionH1::AllocR()
+{
+   const Table& elem_dof_ho = fes_ho.GetElementToDofTable();
+   const Table& elem_dof_lor = fes_lor.GetElementToDofTable();
+   const int ndof_ho = fes_ho.GetNDofs();
+   const int ndof_lor = fes_lor.GetNDofs();
+
+   Table dof_elem_lor;
+   Transpose(elem_dof_lor, dof_elem_lor, ndof_lor);
+
+   Mesh* mesh_lor = fes_lor.GetMesh();
+   const CoarseFineTransformations& cf_tr = mesh_lor->GetRefinementTransforms();
+
+   // mfem::Mult but uses ho2lor to map HO elements to LOR elements
+   const int* elem_dof_hoI = elem_dof_ho.GetI();
+   const int* elem_dof_hoJ = elem_dof_ho.GetJ();
+   const int* dof_elem_lorI = dof_elem_lor.GetI();
+   const int* dof_elem_lorJ = dof_elem_lor.GetJ();
+
+   Array<int> I(ndof_lor + 1);
+
+   // figure out the size of J
+   Array<int> dof_used_ho;
+   dof_used_ho.SetSize(ndof_ho, -1);
+
+   int sizeJ = 0;
+   for (int ilor = 0; ilor < ndof_lor; ++ilor)
+   {
+      for (int jlor = dof_elem_lorI[ilor]; jlor < dof_elem_lorI[ilor + 1]; ++jlor)
+      {
+         int el_lor = dof_elem_lorJ[jlor];
+         int iho = cf_tr.embeddings[el_lor].parent;
+         for (int jho = elem_dof_hoI[iho]; jho < elem_dof_hoI[iho + 1]; ++jho)
+         {
+            int dof_ho = elem_dof_hoJ[jho];
+            if (dof_used_ho[dof_ho] != ilor)
+            {
+               dof_used_ho[dof_ho] = ilor;
+               ++sizeJ;
+            }
+         }
+      }
+   }
+
+   // initialize dof_ho_dof_lor
+   Table dof_lor_dof_ho;
+   dof_lor_dof_ho.SetDims(ndof_lor, sizeJ);
+
+   for (int i = 0; i < ndof_ho; ++i)
+   {
+      dof_used_ho[i] = -1;
+   }
+
+   // set values of J
+   int* dof_dofI = dof_lor_dof_ho.GetI();
+   int* dof_dofJ = dof_lor_dof_ho.GetJ();
+   sizeJ = 0;
+   for (int ilor = 0; ilor < ndof_lor; ++ilor)
+   {
+      dof_dofI[ilor] = sizeJ;
+      for (int jlor = dof_elem_lorI[ilor]; jlor < dof_elem_lorI[ilor + 1]; ++jlor)
+      {
+         int el_lor = dof_elem_lorJ[jlor];
+         int iho = cf_tr.embeddings[el_lor].parent;
+         for (int jho = elem_dof_hoI[iho]; jho < elem_dof_hoI[iho + 1]; ++jho)
+         {
+            int dof_ho = elem_dof_hoJ[jho];
+            if (dof_used_ho[dof_ho] != ilor)
+            {
+               dof_used_ho[dof_ho] = ilor;
+               dof_dofJ[sizeJ] = dof_ho;
+               ++sizeJ;
+            }
+         }
+      }
+   }
+
+   dof_lor_dof_ho.SortRows();
+   double* data = Memory<double>(dof_dofI[ndof_lor]);
+
+   R = new SparseMatrix(dof_dofI, dof_dofJ, data, ndof_lor, ndof_ho,
+                        true, true, true);
+   *R = 0.0;
+
+   dof_lor_dof_ho.LoseData();
+}
+
+
+void L2ProjectionH1GridTransfer::L2ProjectionH1::BuildP() const
+{
+   int ndof_ho = fes_ho.GetNDofs();
+   int ndof_lor = fes_lor.GetNDofs();
+
+   // Compute P = (R^T M_LH)^(-1) M_LH^T column by column (CG solve)
+   SparseMatrix RTxM_LH = *TransposeMult(*R, M_LH);
+   DSmoother Ds(RTxM_LH);
+   P = new DenseMatrix(ndof_ho, ndof_lor);
```

Comment:
I don't think the approach of having `P` as a globally dense matrix is advisable. This will severely limit the problem size for which this is applicable.

A more scalable approach would involve a (preconditioned) CG solve every time `P` or `P^T` is applied (rather than constructed).

Reply:
Agreed.  This has been implemented in the latest commit.